app/src/main/java/com/example/gotproject/MainActivity.kt

```diff
@@ -23,39 +19,36 @@ class MainActivity : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+
         recyclerview_characters = findViewById<RecyclerView>(R.id.recyclerview_characters)
 
         recyclerview_characters.setHasFixedSize(true)
         linearLayoutManager =
             LinearLayoutManager(this)
         recyclerview_characters.layoutManager = linearLayoutManager
-        getMyData()
+
+        initViewModel()
+        initMainViewModel()
+    }
+
+    private fun initViewModel() {
+        recyclerview_characters.apply {
+            linearLayoutManager = LinearLayoutManager(this@MainActivity)
+
+            myAdapter = Adapter()
+        }
     }
 
-    private fun getMyData() {
-        val retrofitBuilder = Retrofit.Builder()
-            .addConverterFactory(GsonConverterFactory.create())
-            .baseUrl(BASE_URL)
-            .build()
-            .create(ApiInterface::class.java)
-
-        val retrofitData = retrofitBuilder.getCharactersData()
-
-        retrofitData.enqueue(object : Callback<List<DataCharactersItem>?> {
-            override fun onResponse(
-                call: Call<List<DataCharactersItem>?>,
-                response: Response<List<DataCharactersItem>?>
-            ) {
-                val responseBody = response.body()!!
-
-                myAdapter = Adapter(baseContext, responseBody)
-                myAdapter.notifyDataSetChanged()
-                recyclerview_characters.adapter = myAdapter
-            }
-
-            override fun onFailure(call: Call<List<DataCharactersItem>?>, t: Throwable) {
-                Log.d("MainActivity", "onFailure" + t.message)
-            }
+    private fun initMainViewModel() {
+        val viewModel = ViewModelProvider(this).get(MainActivityViewModel::class.java)
+        viewModel.getAllDataItemsList().observe(this, Observer<List<DataCharactersItem>> {
+            myAdapter.setListData(it)
+            myAdapter.notifyDataSetChanged()
         })
+
+        viewModel.makeApiCall()
```

Comment:
I recommend doing this call in the ViewModel's init block instead. Since the UI is destroyed and recreated on every configuration change, the ViewModel survives. This means you will only make the API call once, when the ViewModel is created.

Reply:
Moving makeApiCall() in init ViewModel